---
title: "Systems Genomics data analysis pipeline"
---

# install libraries
```{r}
#install.packages(c("tidyverse","ggrepel","BiocManager","pbapply","gplots","msigdbr"))
#BiocManager::install(c("biomaRt","sva","DESeq2","edgeR","tximport"))
#install.packages(c("pheatmap","RColorBrewer","devEMF", "plotly"))
#BiocManager::install(c("svglite", "fgsea"))
#install.packages("stringi", dependencies=TRUE)
#install.packages("viridis")
```

# load libraries
```{r, include=FALSE, }
libraries = c("tidyverse","ggrepel","BiocManager","pbapply","gplots","msigdbr", "biomaRt","sva","DESeq2","edgeR", "dplyr", "biomaRt", "tximport", "ggplot2", "pheatmap", "RColorBrewer", "svglite","devEMF")
lapply(libraries, library, character.only=TRUE)

library(stringi)
library(viridis)
```

```{r}
#pwd <- "/Users/david/Library/CloudStorage/OneDrive-Persönlich/Master ETH/Courses/11_Systems Genomics/project local repository/data"
#pwd2 <- "/Users/david/Library/CloudStorage/OneDrive-Persönlich/Master ETH/Courses/11_Systems Genomics/project local repository"
pwd <- "C:/Users/nicks/Desktop/Systems_Genomics/GitHub 2.0/SystemsGenomics2023/data"
pwd2 <- "C:/Users/nicks/Desktop/Systems_Genomics/GitHub 2.0/SystemsGenomics2023"

# pwd <- "/Users/valentin/Documents/ETH/Master Biotechnologie/Lectures/Sytems_Genomics/Local_git_hub_repository/data"
```

# Nextera trimmed

## Import of data & comparison initial transcriptome analysis
### Import Gene expression quantification (p. 41)
```{r, warning=FALSE}
setwd(pwd)
samples <- list.files("rsem_genes_results_trimmed")

expr <- sapply(samples, function(sample){
  file <- paste0("rsem_genes_results_trimmed/", sample)
  #print(file)
  quant <- read.csv(file, sep="\t", header=T)
  tpm <- setNames(quant$TPM, quant$gene_id)
  return(tpm)
})
SRR_codes <- colnames(expr)
SRR_codes <- gsub("\\.genes\\.results", "", SRR_codes)
colnames(expr) <- SRR_codes
```



### Import metadata for mandatory (meta1) and optional (meta2) dataset (p. 41)
```{r, warning=FALSE}
setwd(pwd)

meta1 <- read.csv("SRR_Acc_List.txt", sep="\t", header=T) %>%
  inner_join(read.csv("PRJNA946653 Metadata.txt", header=T),
    by = c("Name" = "Run"),
    suffix = c("",".y")) 

meta2 <- read.csv("SRR_Acc_List.txt", sep="\t", header=T) %>%
  inner_join(read.csv("PRJNA875066 Metadata.txt", header=T),
    by = c("Name" = "Run"),
    suffix = c("",".y"))
# print(setdiff(colnames(meta1),colnames(meta2)))
meta1_filtered <- subset(meta1, select = c("Name", "experiment_group", "sex", "treatment")) #Compare this to dendrogram

meta1_filtered$sex <- as.factor(meta1_filtered$sex)
meta1_filtered$treatment <- as.factor(meta1_filtered$treatment)
meta1_filtered$experiment_group <- as.factor(meta1_filtered$experiment_group)

meta1_aDR <- meta1_filtered %>% filter(treatment == "AL" | treatment == "aDR" )
meta1_YMP <- meta1_filtered %>% filter(treatment == "YMP" | treatment == "PBS" )

summary(meta1_filtered$sex)
summary(meta1_filtered$treatment)
#expr <- expr[,meta$Name] #to make sure the columns of the expression matrix are in the same order as rows in the metadata
```


### Get more more information about genes with biomaRt (p. 42)
```{r}
#listDatasets(useEnsembl(biomart = "ensembl")) %>% filter(dataset == "mmusculus_gene_ensembl") #Is this the correct one?

library(biomaRt)

ensembl <- useEnsembl(biomart = "ensembl", dataset = "mmusculus_gene_ensembl")
meta_genes <- getBM(attributes = c("ensembl_gene_id",
                                   "ensembl_gene_id_version",
                                   "ensembl_transcript_id_version",
                                   "mgi_symbol", # appearantly necessary for DEseq
                                   "description",
                                   "chromosome_name",
                                   "start_position",
                                   "end_position",
                                   "strand"),
                    filters = "ensembl_gene_id_version",
                    values = rownames(expr),
                    mart = ensembl) %>%
right_join(data.frame(ensembl_gene_id_version = rownames(expr)), by = "ensembl_gene_id_version") %>% distinct(ensembl_gene_id_version, .keep_all = TRUE)

#expr_test <- expr[meta_genes$ensembl_transcript_id_version,]
```



### Comparison of transcriptome profiles 
#### Histograms, filtering unexpressed genes (p. 43-45)
```{r}
# dim(expr) # 56884 annotated genes and 15 samples
avg_expr <- rowMeans(expr)
#layout(matrix(1:2, nrow=1))
hist(avg_expr)
hist(log10(avg_expr + 0.01), main = "Average gene expression (before filtering)", xlab = "log10(Average gene expression [TPM])", ylab = "Number of genes")


ggplot(data.frame(avg_expr), aes(x=avg_expr)) +
  geom_histogram(bins = 50) +
  scale_x_continuous(breaks = c(0,1,10,100,1000,10000,100000), trans="log1p", expand=c(0,0)) +
  scale_y_continuous(breaks = c(0,1), expand=c(0,0), trans="log1p") +
  theme_minimal() #log transformation of the y axis, visualization purposes

num_det <- rowSums(expr > 0) #check in how many samples each gene is detected.
hist(num_det)
#head(num_det)


# get rid of genes that are unexpressed or extremely lowly expressed
expressed <- rowMeans(expr > 0) >= 0.5 | rowMeans(expr) >= 1

meta_genes$expressed <- expressed

head(expressed[meta_genes$expressed]) #expressed is a vector of booleans
summary(expressed[meta_genes$expressed])

#expr <- expr[which(expressed),] #I want to keep the whole expr
#dim(expr[meta_genes$expressed,]) # 25649    15 --> now only 25'649 genes left

avg_expr <- rowMeans(expr[meta_genes$expressed,]) #now after the filtering
#layout(matrix(1:2, nrow=1))
hist(avg_expr)
hist(log10(avg_expr + 0.01), main = "Average gene expression (after filtering)", xlab = "log10(Average gene expression [TPM])", ylab = "Number of genes")
```
#### PCC & SCC (correlation coefficients, cluster dendorgrams, heatmap of samples) (p. 46)
```{r, fig.width=9}
#From here you need meta_genes
#meta_genes$expressed is expressed
#meta_genes$expressed

corr_pearson <- cor(log1p(expr[meta_genes$expressed,]))
corr_spearman <- cor(expr[meta_genes$expressed,], method = "spearman")

#pheatmap(corr_pearson)
plot <- pheatmap((corr_spearman), color = colorRampPalette(c((brewer.pal(n = 9, name = "YlGnBu"))))(100),  # custom color palette YlGnBu
         display_numbers = T,
         number_color = "black",
         number_format = "%.2f")
plot
#ggsave("spearman correlation of treated mice.svg",plot = plot, width = 9, height = 6, path = "/Users/david/Library/CloudStorage/OneDrive-Persönlich/Master ETH/Courses/11_Systems Genomics/project local repository/plots")



#Dendrograms
#hcl_pearson <- hclust(as.dist(1 - corr_pearson))
#hcl_spearman <- hclust(as.dist(1 - corr_spearman))
#layout(matrix(1:2,nrow=1))
#plot(hcl_pearson)
#plot(hcl_spearman)

```

#### PCA
```{r}
pca <- prcomp(log1p(t(expr[meta_genes$expressed,])), center = TRUE,scale.=TRUE)
variances <- pca$sdev^2

variance_proportion <- variances / sum(variances)
cumulative_variance <- cumsum(variance_proportion)


plot <- ggplot(data.frame(pca$x, meta1)) +
#geom_point(aes(x = PC1, y = PC2, color = Name, shape = treatment),size=5)
geom_point(aes(x = PC1, y = PC2, shape = treatment, color = treatment),size=5) +
ggtitle("PCA of mice cerebellum samples") + theme_minimal() + 
  xlab(str_glue("PC1, variance: ",round(variance_proportion[1], 3))) +
  ylab(str_glue("PC2, variance: ",round(variance_proportion[2], 3))) +
  theme(plot.title = element_text(hjust = 0.5, size=16),
        axis.title.y = element_text(size = 14),
        axis.title.x = element_text(size = 14),
        axis.text.y = element_text(size = 11),
        axis.text.x = element_text(size = 11)) 

ggsave("PCA.png",plot = plot, width = 9, height = 4, path = "/Users/valentin/Documents/ETH/Master Biotechnologie/Lectures/Sytems_Genomics/Local_git_hub_repository/plots")

# plot_ly(
#   x = pca$x[,1],
#   y = pca$x[,2],
#   type = "scatter",
#   text = colnames(expr), 
#   mode = "markers",
#   symbol = meta1_filtered$treatment,
#   marker = list(size = 10)
# ) %>%
#   layout(
#     xaxis = list(title = str_glue("PC1, variance: ", round(variance_proportion[1], 3))),
#     yaxis = list(title = str_glue("PC2, variance: ", round(variance_proportion[2], 3))),
#     hovermode = "closest"
#   )

```
##### batch effect correction? --> probbaly does not make sense as we do not have different layers
```{r, eval=FALSE}
library(sva)
expr_combat <- ComBat_seq(counts = expr,
                          batch = meta1$Name)

corr_spearman_combat <- cor(expr_combat[meta_genes$expressed,], method = "spearman")
hcl_spearman_combat <- hclust(as.dist(1 - corr_spearman_combat))
layout(matrix(1:2,nrow=1))
plot(hcl_spearman_combat, labels = meta1$Name)

pca_combat <- prcomp(log1p(t(expr_combat[meta_genes$expressed,])), center = TRUE, scale. = TRUE)
ggplot(data.frame(pca_combat$x, meta1)) +
  geom_point(aes(x = PC1, y = PC2, color = Layer, shape = Name), size = 5)
```

##### Highly variable genes identification (optional)
```{r, eval=FALSE}
#Highly variable genes identification (optional). Manual, We don't do it.
estimate_variability <- function(expr){
  means <- apply(expr, 1, mean)
  vars <- apply(expr, 1, var)
  cv2 <- vars / means^2
  
  
  minMeanForFit <- unname(median(means[which(cv2 > 0.3)]))
  useForFit <- means >= minMeanForFit
  fit <- glm.fit(x = cbind(a0 = 1, a1tilde = 1/means[useForFit]),
    y = cv2[useForFit],
    family = Gamma(link = "identity"))
  a0 <- unname(fit$coefficients["a0"])
  a1 <- unname(fit$coefficients["a1tilde"])
  
  
  xg <- exp(seq(min(log(means[means>0])), max(log(means)), length.out=1000))
  vfit <- a1/xg + a0
  df <- ncol(expr) - 1
  afit <- a1/means+a0
  varFitRatio <- vars/(afit*means^2)
  pval <- pchisq(varFitRatio*df,df=df,lower.tail=F)
  
  
  res <- data.frame(mean = means,
    var = vars,
    cv2 = cv2,
    useForFit = useForFit,
    pval = pval,
    padj = p.adjust(pval, method="BH"),
    row.names = rownames(expr))
  return(res)
}
```



## DESeq 2 (Differential expression Analysis)
### Getting isoforms information for DeSeq2
```{r, warning=FALSE}
ensembl <- useEnsembl(biomart = "ensembl", dataset = "mmusculus_gene_ensembl")
tx2gene <- getBM(attributes = c("ensembl_transcript_id_version","ensembl_gene_id_version"),
                     filters = "ensembl_gene_id_version",
                     values = rownames(expr),
                     mart = ensembl) %>%
  dplyr::select(ensembl_transcript_id_version, ensembl_gene_id_version)
```

```{r, warning=FALSE}
setwd(pwd)

samples <- list.files("rsem_isoforms_results_trimmed") #Hier sollten isoforms sein! nicht dedup nextera!
filtered_samples <- samples[grep("\\.isoforms\\.results$",samples)]# Remove ".isoforms.results" from all column names
files_YMP <- file.path(paste0("rsem_isoforms_results_trimmed/", filtered_samples[1:6]))
files_aDR <- file.path(paste0("rsem_isoforms_results_trimmed/", filtered_samples[7:15]))
txi_YMP <- tximport(files_YMP, type = "rsem", tx2gene = tx2gene)
txi_aDR <- tximport(files_aDR, type = "rsem", tx2gene = tx2gene)

#expr[,15] <- data.frame(txi_aDR)[,9] # this code recovers the broken dataset from mouse 79
```

### DESeq aDR/AL (slide 57)
```{r}
dds_aDR <- DESeqDataSetFromTximport(txi_aDR, colData = meta1_aDR, design = ~ treatment) #the output we be in gene_ids!
#summary(dds_aDR)
rownames(dds_aDR) %>% length() #gene_id_versions
# txi_aDR, dds_aDR: 56884 genes

dds_filtered_aDR <- dds_aDR[intersect(meta_genes[meta_genes$expressed,2],rownames(dds_aDR)),]
rownames(dds_filtered_aDR) %>% length() #gene_id_versions
# sum(meta_genes$expressed): 25649; dds_filtered_aDR: 25649


dds_filtered_aDR <- DESeq(dds_filtered_aDR, test="LRT", reduced= ~ 1) #If only one covariate, use ~1

res_DESeq2_aDR <- results(dds_filtered_aDR) # 25649 genes, same amount
res_DESeq2_aDR <- data.frame(res_DESeq2_aDR) 
res_DESeq2_aDR$pvalue_BH <- p.adjust(res_DESeq2_aDR$pvalue, method="BH") #Argumentieren wieso BH correction
res_DESeq2_aDR$minuslog10_pvalue_BH <- -log10(res_DESeq2_aDR$pvalue_BH)

sum(is.na(res_DESeq2_aDR$log2FoldChange))
# the meta_genes dataset contains 5962 ensembl_gene_id_version that are not found in expr,
# because we filtered expr based on the criteria rowMeans(expr > 0) >= 0.5 | rowMeans(expr) >= 1

sum(is.na(res_DESeq2_aDR$pvalue)) # 6009
temp_mask = is.na(res_DESeq2_aDR$pvalue) & !is.na(res_DESeq2_aDR$log2FoldChange)
# res_DESeq2_aDR[temp_mask,] # However, for some reason DESeq2 fucks up the pvalue for these 47 genes
res_DESeq2_aDR_fitered <- res_DESeq2_aDR[!is.na(res_DESeq2_aDR$pvalue_BH),]
res_DESeq2_aDR_DEG <- res_DESeq2_aDR_fitered[res_DESeq2_aDR_fitered$pvalue_BH < 0.05,]#filter to only have significant data, include FC?
DEG_aDR <- rownames(res_DESeq2_aDR_DEG)


DEG_comaprison <-  data_frame("trim status" ="trimmed", "Rejuvenation approach"="aDR",DEGs = length(DEG_aDR))
```

#### Volcano Plot
```{r, fig.width=7, fig.height=7}
#plot(res_DESeq2_aDR$log2FoldChange, p.adjust(res_DESeq2_aDR$pvalue, method = "BH"))
#plot(res_DESeq2_aDR$log2FoldChange, -log10(p.adjust(res_DESeq2_aDR$pvalue)), method = "BH")

#BiocManager::install("EnhancedVolcano")
library(EnhancedVolcano)
FCcutoff = 1

plot <- EnhancedVolcano(res_DESeq2_aDR_fitered,
                lab = meta_genes[meta_genes$ensembl_gene_id_version %in% 
                                   rownames(res_DESeq2_aDR_fitered), "mgi_symbol"],
                labSize = 5,
                x = "log2FoldChange",
                y = "pvalue_BH",
                title = "DEG of DESeq2 aDR/AL",
                subtitle = "",
                pCutoff = 0.05,
                FCcutoff = 1,
                pointSize = 3,
                selectLab = res_DESeq2_aDR_fitered$log2FoldChange < FCcutoff & 
                  res_DESeq2_aDR_fitered$log2FoldChange < -FCcutoff,
                xlim = c(-6,6),
                ylim = c(0, 10),
                drawConnectors = TRUE,
                directionConnectors = 'both',
                arrowheads = FALSE,
                boxedLabels = FALSE,
                legendPosition = "none"
)
plot
ggsave("Volcano plot aDR_AL DESeq2.png",plot = plot, width = 8, height = 8, path = "/Users/valentin/Documents/ETH/Master Biotechnologie/Lectures/Sytems_Genomics/Local_git_hub_repository/plots")
```

#### Grouping of identified genes (p. 59)
```{r}
expr_aDR <- expr[,7:15]
treatment_aDR <- unique(meta1_aDR$treatment)
avg_sc_expr_aDR <- sapply(treatment_aDR, function(treatment)
  #rowMeans(expr_aDR[,which(meta1_aDR$treatment == treatment)])) #V1: aDR, V2: AL
  rowMeans(t(scale(t(expr_aDR[,which(meta1_aDR$treatment == treatment)]))))) 
colnames(avg_sc_expr_aDR) <- treatment_aDR

head(avg_sc_expr_aDR)
avg_sc_expr_aDR <- data.frame(avg_sc_expr_aDR)
```

```{r}
max_treatment_DEG_aDR <- setNames(colnames(avg_sc_expr_aDR)[apply(avg_sc_expr_aDR[DEG_aDR,], 1, which.max)],DEG_aDR)
avg_expr_DEG_list_aDR <- tapply(names(max_treatment_DEG_aDR), max_treatment_DEG_aDR, function(x) avg_sc_expr_aDR[x,])

## centered but not scaled
# scaled_expr_DEG_list_aDR <- lapply(avg_expr_DEG_list_aDR, function(x) t(scale(t(x),scale = FALSE))) # scaling converts all of the genes into one value instead of a normalized distribtion
# layout(matrix(1:2, nrow = 1, byrow = T))
# par(mar=c(3,3,3,3))
# for(treatment in names(scaled_expr_DEG_list_aDR))
# boxplot(scaled_expr_DEG_list_aDR[[treatment]],
# main = paste0(treatment, " (", nrow(avg_expr_DEG_list_aDR[[treatment]]),")"))

# centered and scaled

layout(matrix(1:2, nrow = 1, byrow = T))
par(mar=c(3,3,3,3))
for(treatment in names(avg_expr_DEG_list_aDR))
  boxplot(avg_expr_DEG_list_aDR[[treatment]],
          main = paste0(treatment, " (", nrow(avg_expr_DEG_list_aDR[[treatment]]),")"),
          ylim = c(-3.5e-15, 3.5e-15))
```

#### hierarchical clustering of aDR GENES (MICE is below in a new chunk)
```{r}
#corr_DEG_mice <- cor(expr_aDR[DEG_aDR,], method = "spearman")
corr_DEG_genes <- cor(t(expr_aDR[DEG_aDR,]), method = "spearman")

library(pheatmap)
pheatmap(corr_DEG_genes)
hcl_DEG_genes <- hclust(as.dist(1 - corr_DEG_genes), method = "complete")
plot(hcl_DEG_genes, labels = FALSE)

#Virdis color palette
#install.packages("viridis")
library(viridis)

setwd(paste0(pwd2, "/plots"))
png("Heatmap_genes_aDR.png", width = 7, height = 4.5, units = "in", res = 300)
heatmap.2(corr_DEG_genes, Rowv = as.dendrogram(hcl_DEG_genes), Colv = as.dendrogram(hcl_DEG_genes),
trace = "none", scale = "none", labRow = NA, labCol = NA, col = viridis)
dev.off()

cl_DEG_genes <- cutree(hcl_DEG_genes, k = 2) #k can be adjusted
heatmap.2(corr_DEG_genes, Rowv = as.dendrogram(hcl_DEG_genes), Colv = as.dendrogram(hcl_DEG_genes),
trace = "none", scale = "none", labRow = NA, labCol = NA, col = viridis,
ColSideColors = rainbow(15)[cl_DEG_genes]) 
```
#### hierarchical clustering of aDR MICE (GENES is above)
```{r}
corr_DEG_mice <- cor(expr_aDR[DEG_aDR,], method = "spearman")
#corr_DEG_genes <- cor(t(expr_aDR[DEG_aDR,]), method = "spearman")

library(pheatmap)
pheatmap(corr_DEG_mice)
hcl_DEG_mice <- hclust(as.dist(1 - corr_DEG_mice), method = "complete")
plot(hcl_DEG_mice, labels = FALSE)

#Virdis color palette
#install.packages("viridis")
library(viridis)
heatplot <- heatmap.2(corr_DEG_mice, Rowv = as.dendrogram(hcl_DEG_mice), Colv = as.dendrogram(hcl_DEG_mice),
trace = "none", scale = "none", labRow = NA, labCol = NA, col = viridis::magma(100))
ggsave("Heatmap_samles_aDR.png",plot = heatplot, width = 5, height = 5, path = paste0(pwd2, "/plots"))

cl_DEG_mice <- cutree(hcl_DEG_mice, k = 4) #k can be adjusted
heatmap.2(corr_DEG_mice, Rowv = as.dendrogram(hcl_DEG_mice), Colv = as.dendrogram(hcl_DEG_mice),
trace = "none", scale = "none", labRow = NA, labCol = NA, col = viridis::magma(100),
ColSideColors = rainbow(15)[cl_DEG_mice]) 
```

#####save heatmap

```{r}
setwd(paste0(pwd2, "/plots"))
png("Heatmap_samles_aDR.png", width = 7, height = 4.5, units = "in", res = 300)
heatmap_obj <- heatmap.2(corr_DEG_mice, 
                        Rowv = as.dendrogram(hcl_DEG_mice), 
                        Colv = as.dendrogram(hcl_DEG_mice),
                        trace = "none", 
                        scale = "none", 
                        labRow = NA, 
                        labCol = NA, 
                        col = viridis::magma(100))
dev.off()
```


#### 3-6 Making sense of the genes (page 66)
```{r}
#aDR

#Convert list of DEG's into a format that you can upload into DAVID.
#Think of what cluster of genes we want to analyze!
#write.table(meta_genes[meta_genes$ensembl_gene_id_version %in% names(which(cl_DEG==2)), "ensembl_gene"], file = "genes_C2.txt",quote = F, row.names = F, col.names = F) #DEG's

#Convert DEG gene id versions to gene id (this is what DAVID wants)
# Remove everything after the dot

setwd(pwd)

DEG_aDR_gene_id <- sub("\\..*", "", DEG_aDR)


write.table(DEG_aDR_gene_id,
file = "genes_C2.txt",
quote = F, row.names = F, col.names = F) #DEG's, dies ist der Code der meiner Meinung nach Sinn macht.

write.table(meta_genes[meta_genes$expressed, "ensembl_gene_id"],
file = "genes_expressed.txt",
quote = F, row.names = F, col.names = F) #All expressed genes (background gene panel)
```

#### Gene Ontology enrichment GORILLA (page 70)
```{r}
# webtool: https://cbl-gorilla.cs.technion.ac.il/
# using two unranked lists of genes
# Results: https://cbl-gorilla.cs.technion.ac.il/GOrilla/q5z11jdw/GOResults.html#genes_info
setwd("/Users/david/Library/CloudStorage/OneDrive-Persönlich/Master ETH/Courses/11_Systems Genomics/project local repository")

GO_aDR <- read.csv("GOrilla/GO_aDR.csv", header = TRUE, sep = ";")
GO_aDR <- GO_aDR %>%
  mutate_at(vars(N, B, n, b, Enrichment), as.numeric)
GO_aDR$Enrichment <- (GO_aDR$b/GO_aDR$n) / (GO_aDR$B/GO_aDR$N)
GO_aDR$Genes <- sub("^\\[|\\]$", "", GO_aDR$Genes)


myelination <- unlist(str_split(GO_aDR$Genes[1], ","))
myelination <- gsub("-.*", "", myelination)
myelination <- gsub(" ", "", myelination)
myelin_meta <- meta_genes[meta_genes$mgi_symbol %in% myelination ,]
myelin_expr <- avg_sc_expr_aDR[rownames(expr) %in% myelin_meta$ensembl_gene_id_version,]
myelin_expr$mgi_symbol <- myelin_meta$mgi_symbol

pos_cellular_process <- unlist(str_split(GO_aDR$Genes[4], ","))
pos_cellular_process <- gsub("-.*", "", pos_cellular_process)
pos_cellular_process <- gsub(" ", "", pos_cellular_process)
pos_cellular_process <- c(pos_cellular_process, "Prxl2c", "Niban2")
cell_proc_meta <- meta_genes[meta_genes$mgi_symbol %in% pos_cellular_process ,]
cell_proc_expr <- avg_sc_expr_aDR[rownames(expr) %in% cell_proc_meta$ensembl_gene_id_version,]
cell_proc_expr$mgi_symbol <- cell_proc_meta$mgi_symbol

pos_reg_met <- unlist(str_split(GO_aDR$Genes[9], ","))
pos_reg_met <- gsub("-.*", "", pos_reg_met)
pos_reg_met <- gsub(" ", "", pos_reg_met)
reg_met_meta <- meta_genes[meta_genes$mgi_symbol %in% pos_reg_met ,]
reg_met_expr <- avg_sc_expr_aDR[rownames(expr) %in% reg_met_meta$ensembl_gene_id_version,]

length(intersect(pos_reg_met,pos_cellular_process))# lol, same genes.

#meta_genes[!meta_genes$mgi_symbol %in% pos_cellular_process ,]
#pos_cellular_process[!pos_cellular_process %in%  meta_genes$mgi_symbol]


#meta_genes[meta_genes$ensembl_gene_id == "ENSMUSG00000026796",]
#res_DESeq2_aDR["ENSMUSG00000021482.11",]

max_treatment_DEG_aDR <- setNames(colnames(avg_sc_expr_aDR)[apply(avg_sc_expr_aDR[DEG_aDR,], 1, which.max)],DEG_aDR)
avg_expr_DEG_list_aDR <- tapply(names(max_treatment_DEG_aDR), max_treatment_DEG_aDR, function(x) avg_sc_expr_aDR[x,])

myelin_list_aDR <- lapply(avg_expr_DEG_list_aDR, function(df) df[rownames(df) %in% myelin_meta$ensembl_gene_id_version,])
pos_cell_list_aDR <- lapply(avg_expr_DEG_list_aDR, function(df) df[rownames(df) %in% cell_proc_meta$ensembl_gene_id_version,])

layout(matrix(1:2, nrow = 1, byrow = T))
par(mar=c(3,3,3,3))
for(treatment in names(myelin_list_aDR))
  boxplot(myelin_list_aDR[[treatment]],
          main = paste0(treatment, " (", nrow(myelin_list_aDR[[treatment]]),")"),
          ylim = c(-6e-16, 6e-16))

layout(matrix(1:2, nrow = 1, byrow = T))
par(mar=c(3,3,3,3))
for(treatment in names(pos_cell_list_aDR))
  boxplot(pos_cell_list_aDR[[treatment]],
          main = paste0(treatment, " (", nrow(pos_cell_list_aDR[[treatment]]),")"),
          ylim = c(-6e-15, 6e-15))
```


```{r}
# myelin_expr contains expr data for all enriched genes for AD and aDR and another column with the mgi_symbol
myelin_expr_long <- myelin_expr %>%
  gather(key = "treatment", value = "Expression", -mgi_symbol) # reshape the df
myelin_expr_long$Expression <- round(myelin_expr_long$Expression * 10e15, digits = 3)

# Plot
grouping_plot <- ggplot(myelin_expr_long, aes(x = treatment, y = Expression, group = mgi_symbol, color = mgi_symbol)) +
  geom_line() +
  geom_point() +
  labs(title = "Process gene enrichment\nin aDR: Myelination",
       x = "Treatment",
       y = "Normalized expression") +
  theme_minimal() +
  theme(plot.title = element_text(hjust = 0.5))
grouping_plot
ggsave("Grouping_GO_Myelination.png",plot = grouping_plot, width = 3, height = 3, path = paste0(pwd2, "/plots"))


cell_proc_expr_long <- cell_proc_expr %>%
  gather(key = "treatment", value = "Expression", -mgi_symbol) # reshape the df
cell_proc_expr_long$Expression <- round(cell_proc_expr_long$Expression * 10e14, digits = 3)

grouping_plot <- ggplot(cell_proc_expr_long, aes(x = treatment, y = Expression, group = mgi_symbol, color = mgi_symbol)) +
  geom_line() +
  geom_point() +
  labs(title = "Process gene enrichment\nin aDR: Cellular processes",
       x = "Treatment",
       y = "Normalized expression") +
  theme_minimal() +
  theme(plot.title = element_text(hjust = 0.5))
grouping_plot
ggsave("Grouping_GO_Cellular_processes.png",plot = grouping_plot, width = 5.5, height = 4, path = paste0(pwd2, "/plots"))



```
#### Gene Ontology enrichment DAVID aDR (page 70)
```{r}
# webtool: https://cbl-gorilla.cs.technion.ac.il/
# using two unranked lists of genes
# Results: https://cbl-gorilla.cs.technion.ac.il/GOrilla/q5z11jdw/GOResults.html#genes_info
setwd(pwd2)
#Biological rhythms, 6 genes
DA_aDR <- read.table("DAVID/DAVID aDR biological rhythms.txt", header = TRUE, sep = "\t")
head(DA_aDR)
DA_aDR$gene_id <- rownames(DA_aDR)
DA_aDR <- data.frame(DA_aDR)

DA_aDR_gene_ids <- meta_genes[meta_genes$ensembl_gene_id %in% DA_aDR$gene_id,]
#DA_aDR_expr <- avg_sc_expr_aDR[DA_aDR_gene_ids$ensembl_gene_id,]
DA_aDR_expr <- avg_sc_expr_aDR[DA_aDR_gene_ids$ensembl_gene_id_version,] 
DA_aDR_expr$mgi_symbol <- meta_genes[meta_genes$ensembl_gene_id_version %in% rownames(DA_aDR_expr),"mgi_symbol"]

#DA_aDR_expr <- left_join(DA_aDR_expr, DA_aDR_gene_ids)
head(DA_aDR_expr)


#Create a nice, corresponding graph
# myelin_expr contains expr data for all enriched genes for AD and aDR and another column with the mgi_symbol
DA_aDR_expr_long <- DA_aDR_expr %>%
  gather(key = "treatment", value = "Expression", -mgi_symbol) # reshape the df
DA_aDR_expr_long$Expression <- round(DA_aDR_expr_long$Expression * 10e16,digits=3)
# Plot
grouping_plot <- ggplot(DA_aDR_expr_long, aes(x = treatment, y = Expression, group = mgi_symbol, color = mgi_symbol)) +
  geom_line() +
  geom_point() +
  labs(title = "Process gene enrichment\nin aDR: Biological rhythms",
       x = "Treatment",
       y = "Normalized expression") +
  theme_minimal() +
  theme(plot.title = element_text(hjust = 0.5))
grouping_plot
ggsave("Grouping_DAVID_biological_rhythms.png",plot = grouping_plot, width = 3.5, height = 3, path = paste0(pwd2,"/plots"))


```

```{r}
#Testing chat gpt code
# Assuming your gene expression data is stored in a data frame called 'gene_expression'
# with row names as gene names and columns 'condition_a' and 'condition_b'
#install.packages("reshape2")
library(ggplot2)
library(tidyr)


# Assuming your gene expression data is stored in a data frame called 'gene_expression'
# with row names as gene names and columns 'condition_a' and 'condition_b'

library(ggplot2)
library(tidyr)

# Sample data
set.seed(123)
gene_names <- letters[1:10]
gene_expression <- data.frame(
  condition_a = rnorm(10, mean = 10, sd = 2),
  condition_b = rnorm(10, mean = 12, sd = 2)
)
rownames(gene_expression) <- gene_names
gene_expression$name <- rownames(gene_expression)


# Convert data to long format using tidyr::pivot_longer
gene_expression_long <- pivot_longer(gene_expression, cols = c(condition_a, condition_b), names_to = "Condition", values_to = "Expression")

# Create a scatter plot with connecting lines
ggplot(gene_expression_long, aes(x = Condition, y = Expression, group = name, color = name)) +
  geom_point() +
  geom_line() +
  labs(title = "Gene Expression in Conditions A and B",
       x = "Conditions", y = "Expression Level") +
  theme_minimal()



```

```{r}


GO_aDR <- GO_aDR %>%
  mutate_at(vars(N, B, n, b, Enrichment), as.numeric)
GO_aDR$Enrichment <- (GO_aDR$b/GO_aDR$n) / (GO_aDR$B/GO_aDR$N)
GO_aDR$Genes <- sub("^\\[|\\]$", "", GO_aDR$Genes)


myelination <- unlist(str_split(GO_aDR$Genes[1], ","))
myelination <- gsub("-.*", "", myelination)
myelination <- gsub(" ", "", myelination)
myelin_meta <- meta_genes[meta_genes$mgi_symbol %in% myelination ,]
myelin_expr <- avg_sc_expr_aDR[rownames(expr) %in% myelin_meta$ensembl_gene_id_version,]

pos_cellular_process <- unlist(str_split(GO_aDR$Genes[4], ","))
pos_cellular_process <- gsub("-.*", "", pos_cellular_process)
pos_cellular_process <- gsub(" ", "", pos_cellular_process)
pos_cellular_process <- c(pos_cellular_process, "Prxl2c", "Niban2")
cell_proc_meta <- meta_genes[meta_genes$mgi_symbol %in% pos_cellular_process ,]
cell_proc_expr <- avg_sc_expr_aDR[rownames(expr) %in% cell_proc_meta$ensembl_gene_id_version,]

pos_reg_met <- unlist(str_split(GO_aDR$Genes[9], ","))
pos_reg_met <- gsub("-.*", "", pos_reg_met)
pos_reg_met <- gsub(" ", "", pos_reg_met)
reg_met_meta <- meta_genes[meta_genes$mgi_symbol %in% pos_reg_met ,]
reg_met_expr <- avg_sc_expr_aDR[rownames(expr) %in% reg_met_meta$ensembl_gene_id_version,]

length(intersect(pos_reg_met,pos_cellular_process))# lol, same genes.

#meta_genes[!meta_genes$mgi_symbol %in% pos_cellular_process ,]
#pos_cellular_process[!pos_cellular_process %in%  meta_genes$mgi_symbol]


#meta_genes[meta_genes$ensembl_gene_id == "ENSMUSG00000026796",]
#res_DESeq2_aDR["ENSMUSG00000021482.11",]

max_treatment_DEG_aDR <- setNames(colnames(avg_sc_expr_aDR)[apply(avg_sc_expr_aDR[DEG_aDR,], 1, which.max)],DEG_aDR)
avg_expr_DEG_list_aDR <- tapply(names(max_treatment_DEG_aDR), max_treatment_DEG_aDR, function(x) avg_sc_expr_aDR[x,])

myelin_list_aDR <- lapply(avg_expr_DEG_list_aDR, function(df) df[rownames(df) %in% myelin_meta$ensembl_gene_id_version,])
pos_cell_list_aDR <- lapply(avg_expr_DEG_list_aDR, function(df) df[rownames(df) %in% cell_proc_meta$ensembl_gene_id_version,])

layout(matrix(1:2, nrow = 1, byrow = T))
par(mar=c(3,3,3,3))
for(treatment in names(myelin_list_aDR))
  boxplot(myelin_list_aDR[[treatment]],
          main = paste0(treatment, " (", nrow(myelin_list_aDR[[treatment]]),")"),
          ylim = c(-6e-16, 6e-16))

layout(matrix(1:2, nrow = 1, byrow = T))
par(mar=c(3,3,3,3))
for(treatment in names(pos_cell_list_aDR))
  boxplot(pos_cell_list_aDR[[treatment]],
          main = paste0(treatment, " (", nrow(pos_cell_list_aDR[[treatment]]),")"),
          ylim = c(-6e-15, 6e-15))


# layout(matrix(1:4, nrow = 2, byrow = T))
# par(mar=c(3,3,3,3))
# for(GOset in c(myelin_list_aDR, pos_cell_list_aDR)) {
#   for(treatment in names(GOset)) {
#     boxplot(GOset[[treatment]],
#             main = paste0(treatment, " (", nrow(GOset[[treatment]]),")"))
#   }
# }
  
```



#### Gene set Enrichment analysis (page 70)
```{r, eval=FALSE}
# sadly no meaningful results

score_aDR <- setNames(sign(res_DESeq2_aDR_DEG$log2FoldChange) * res_DESeq2_aDR_DEG$minuslog10_pvalue_BH, gsub("\\..*","", DEG_aDR))
score_ordered_aDR <- sort(score_aDR, decreasing=TRUE)
#BiocManager::install("fgsea")
library(msigdbr)
genesets_celltype <- msigdbr(species = "Mus musculus", category = "H")
genesets_celltype <- genesets_celltype[!is.na(genesets_celltype$ensembl_gene),] # 32 entries have NA values in ensembl_gene_id
genesets_celltype_list <- tapply(genesets_celltype$ensembl_gene, genesets_celltype$gs_name, list)

# length(intersect(unique(genesets_celltype$ensembl_gene), names(score_ordered_aDR)))

library(fgsea)
fgsea_kegg <- fgsea(pathways = genesets_celltype_list, stats = score_ordered_aDR, minSize = 10, maxSize = 100)
fgsea_kegg
fgsea_kegg[order(NES,decreasing=T),][1:10,1:7]
plotEnrichment(genesets_celltype_list[["MURARO_PANCREAS_DUCTAL_CELL"]],
score_ordered_aDR) + labs(title="hNbML5 GABAergic neurons")
#fgsea_kegg[order(NES,decreasing=F),][1:10,1:7]
```


### 3.7 Other analysis (page 72)

### Cas genes from paper
```{r}
setwd("/Users/david/Library/CloudStorage/OneDrive-Persönlich/Master ETH/Courses/11_Systems Genomics/project local repository")
CAS_genes_paper <- read.table("data/CAS genes.txt", header = TRUE, sep = "\t")
#CAS_genes <- meta_genes[meta_genes$mgi_symbol == CAS_genes$Gene_Symbol,]


#write.table(CAS_genes$Gene_Symbol, file = "CAS_gene_symbol.txt", quote = F, row.names = F, col.names = F)
CAS_genes <- read.table("Cas_genes_ensembl.txt", header = TRUE, sep = "\t")
#CAS_genes <- CAS_genes[ CAS_genes$Symbol %in%  CAS_genes_paper$Gene_Symbol,]

length(unique(CAS_genes$Ensembl.ID))
#setdiff(CAS_genes$Symbol,CAS_genes_paper$Gene_Symbol))

# how many of the CAS genes are in our DEG?
common_CAS_DEG_aDR <- intersect(CAS_genes$Ensembl_ID, gsub("\\..*","", DEG_aDR))
CAS_genes[CAS_genes$Ensembl.ID %in% common_CAS_DEG_aDR,]


```

### Cerebellum Genes
```{r}
setwd("/Users/david/Library/CloudStorage/OneDrive-Persönlich/Master ETH/Courses/11_Systems Genomics/project local repository")
Cereb_paper <- read.table("data/region specific DEG in Cerebellum.txt", header = TRUE, sep = "\t")

#write.table(Cereb_paper$Gene_Symbol, file = "Cereb_paper.txt", quote = F, row.names = F, col.names = F)

Cereb_genes <- read.csv("cereb_genes.csv", header = TRUE, sep = ";")

length(unique(Cereb_genes$Ensembl.ID))
#setdiff(CAS_genes$Symbol,CAS_genes_paper$Gene_Symbol))

# how many of the CAS genes are in our DEG?
common_cereb_DEG_aDR <- intersect(Cereb_genes$Ensembl.ID, gsub("\\..*","", DEG_aDR))
CAS_genes[Cereb_genes$Ensembl.ID %in% common_CAS_DEG_aDR,]
```


### DESeq YMP/PBS (slide 57)
```{r}
dds_YMP <- DESeqDataSetFromTximport(txi_YMP, colData = meta1_YMP, design = ~ treatment) #the output we be in gene_ids!
rownames(dds_YMP) %>% length() #gene_id_versions
# txi_aDR, dds_aDR: 56'884 genes

dds_filtered_YMP <- dds_YMP[intersect(meta_genes[meta_genes$expressed,2],rownames(dds_YMP)),]
rownames(dds_filtered_YMP) %>% length() #gene_id_versions
# sum(meta_genes$expressed): 25232; dds_filtered_aDR: 25649; + 417 genes, intersect broken?


dds_filtered_YMP <- DESeq(dds_filtered_YMP, test="LRT", reduced= ~ 1) #If only one covariate, use ~1

res_DESeq2_YMP <- results(dds_filtered_YMP) # 25649 genes, same amount
res_DESeq2_YMP <- data.frame(res_DESeq2_YMP)
res_DESeq2_YMP$log2FoldChange <- -res_DESeq2_YMP$log2FoldChange
res_DESeq2_YMP$pvalue_BH <- p.adjust(res_DESeq2_YMP$pvalue, method="BH") #Argumentieren wieso BH correction
res_DESeq2_YMP$minuslog10_pvalue_BH <- -log10(res_DESeq2_YMP$pvalue_BH)

sum(is.na(res_DESeq2_YMP$log2FoldChange))
# the meta_genes dataset contains 6702 ensembl_gene_id_version that are not found in expr,
# because we filtered expr based on the criteria rowMeans(expr > 0) >= 0.5 | rowMeans(expr) >= 1

sum(is.na(res_DESeq2_YMP$pvalue)) # 6704
temp_mask = is.na(res_DESeq2_YMP$pvalue) & !is.na(res_DESeq2_YMP$log2FoldChange)
# res_DESeq2_YMP[temp_mask,] # However, for some reason DESeq2 fucks up the pvalue for these 47 genes
res_DESeq2_YMP_fitered <- res_DESeq2_YMP[!is.na(res_DESeq2_YMP$pvalue_BH),]
res_DESeq2_YMP_DEG <- res_DESeq2_YMP_fitered[res_DESeq2_YMP_fitered$pvalue_BH < 0.05,]#filter to only have significant data, include FC?
DEG_YMP <- rownames(res_DESeq2_YMP_DEG)

new_row <- c("trimmed", "YMP", length(DEG_YMP))
DEG_comaprison <- rbind(DEG_comaprison, new_row)
```

#### Volcano Plot
```{r, fig.width=7, fig.height=7}
#plot(res_DESeq2_aDR$log2FoldChange, p.adjust(res_DESeq2_aDR$pvalue, method = "BH"))
#plot(res_DESeq2_aDR$log2FoldChange, -log10(p.adjust(res_DESeq2_aDR$pvalue)), method = "BH")

#BiocManager::install("EnhancedVolcano")
library(EnhancedVolcano)

plot <- EnhancedVolcano(res_DESeq2_YMP_fitered,
                lab = meta_genes[meta_genes$ensembl_gene_id_version %in% rownames(res_DESeq2_YMP_fitered), "mgi_symbol"],
               labSize = 5,
                x = "log2FoldChange",
                y = "pvalue_BH",
                title = "DEG of DESeq2 aDR/AL",
                subtitle = "",
                pCutoff = 0.05,
                FCcutoff = 1,
                pointSize = 3,
                selectLab = res_DESeq2_aDR_fitered$log2FoldChange < FCcutoff & 
                  res_DESeq2_aDR_fitered$log2FoldChange < -FCcutoff,
                xlim = c(-6,6),
                ylim = c(0, 10),
                drawConnectors = TRUE,
                directionConnectors = 'both',
                arrowheads = FALSE,
                boxedLabels = FALSE,
                legendPosition = "none"
)
plot
ggsave("Volcano plot YMP_PBS DESeq2.png",plot = plot, width = 8, height = 8, path = "/Users/valentin/Documents/ETH/Master Biotechnologie/Lectures/Sytems_Genomics/Local_git_hub_repository/plots")

```


#### Grouping of identified genes (p. 59)
```{r}
expr_YMP <- expr[,1:6]
treatment_YMP <- unique(meta1_YMP$treatment)
avg_sc_expr_YMP <- sapply(treatment_YMP, function(treatment)
rowMeans(t(scale(t(expr_YMP[,which(meta1_YMP$treatment == treatment)]))))) #V1: YMP, V2: PBS
colnames(avg_sc_expr_YMP) <- treatment_YMP
head(avg_sc_expr_YMP)

avg_sc_expr_YMP <- data.frame(avg_sc_expr_YMP)
```

```{r}
max_treatment_DEG_YMP <- setNames(colnames(avg_sc_expr_YMP)[apply(avg_sc_expr_YMP[DEG_YMP,], 1, which.max)],DEG_YMP)
avg_expr_DEG_list_YMP <- tapply(names(max_treatment_DEG_YMP), max_treatment_DEG_YMP, function(x) avg_sc_expr_YMP[x,])

layout(matrix(1:2, nrow = 1, byrow = T))
par(mar=c(3,3,3,3))
for(treatment in names(avg_expr_DEG_list_YMP))
boxplot(avg_expr_DEG_list_YMP[[treatment]],
main = paste0(treatment, " (", nrow(avg_expr_DEG_list_YMP[[treatment]]),")"),
ylim = c(-2.5e-15, 2.5e-15))

```

#### hierarchical clustering of YMP GENES (MICE is below in a new chunk)
```{r}
#corr_DEG_mice <- cor(expr_aDR[DEG_aDR,], method = "spearman")
corr_DEG_genes <- cor(t(expr_YMP[DEG_YMP,]), method = "spearman")

library(pheatmap)
pheatmap(corr_DEG_genes)
hcl_DEG_genes <- hclust(as.dist(1 - corr_DEG_genes), method = "complete")
plot(hcl_DEG_genes, labels = FALSE)

#Virdis color palette
#install.packages("viridis")

#setwd(paste0(pwd2, "/plots"))
#png("Heatmap_genes_YMP.png", width = 7, height = 4.5, units = "in", res = 300)
library(viridis)
heatmap.2(corr_DEG_genes, Rowv = as.dendrogram(hcl_DEG_genes), Colv = as.dendrogram(hcl_DEG_genes),
trace = "none", scale = "none", labRow = NA, labCol = NA, col = viridis)
#dev.off()

cl_DEG_genes <- cutree(hcl_DEG_genes, k = 4) #k can be adjusted
heatmap.2(corr_DEG_genes, Rowv = as.dendrogram(hcl_DEG_genes), Colv = as.dendrogram(hcl_DEG_genes),
trace = "none", scale = "none", labRow = NA, labCol = NA, col = viridis,
ColSideColors = rainbow(15)[cl_DEG_genes]) 
```


#### hierarchical clustering of YMP MICE (GENES is above)
```{r}
corr_DEG_mice <- cor(expr_YMP[DEG_YMP,], method = "spearman")
#corr_DEG_genes <- cor(t(expr_aDR[DEG_aDR,]), method = "spearman")

library(pheatmap)
pheatmap(corr_DEG_mice)
hcl_DEG_mice <- hclust(as.dist(1 - corr_DEG_mice), method = "complete")
plot(hcl_DEG_mice, labels = FALSE)

#Virdis color palette
#install.packages("viridis")
library(viridis)
heatmap.2(corr_DEG_mice, Rowv = as.dendrogram(hcl_DEG_mice), Colv = as.dendrogram(hcl_DEG_mice),
trace = "none", scale = "none", labRow = NA, labCol = NA, col = viridis)

cl_DEG_mice <- cutree(hcl_DEG_mice, k = 2) #k can be adjusted
heatmap.2(corr_DEG_mice, Rowv = as.dendrogram(hcl_DEG_mice), Colv = as.dendrogram(hcl_DEG_mice),
trace = "none", scale = "none", labRow = NA, labCol = NA, col = viridis,
ColSideColors = rainbow(15)[cl_DEG_mice]) 
```

<<<<<<< HEAD
#### 3-6 Making sense of the genes YMP (page 66)
=======

```{r}
setwd(paste0(pwd2, "/plots"))
png("Heatmap_samles_YMP.png", width = 7, height = 4.5, units = "in", res = 300)
heatmap.2(corr_DEG_mice,
          Rowv = as.dendrogram(hcl_DEG_mice), 
          Colv = as.dendrogram(hcl_DEG_mice),
          trace = "none",
          scale = "none", 
          labRow = NA, 
          labCol = NA, 
          col = viridis::magma(100))
dev.off()
```


### 3-6 Making sense of the genes (page 66)
>>>>>>> 801d201c6955fef722b0894a20905da9ea7d07c2
```{r}
#YMP

#Convert list of DEG's into a format that you can upload into DAVID.
#Think of what cluster of genes we want to analyze!
#write.table(meta_genes[meta_genes$ensembl_gene_id_version %in% names(which(cl_DEG==2)), "ensembl_gene"], file = "genes_C2.txt",quote = F, row.names = F, col.names = F) #DEG's

#Convert DEG gene id versions to gene id (this is what DAVID wants)
# Remove everything after the dot

setwd(pwd)
getwd()
DEG_YMP_gene_id <- sub("\\..*", "", DEG_YMP)


write.table(DEG_YMP_gene_id, file = "DEG_genes_YMP.txt", quote = F, row.names = F, col.names = F) #DEG's, dies ist der Code der meiner Meinung nach Sinn macht.

# write.table(meta_genes[meta_genes$expressed, "ensembl_gene_id"],
# file = "genes_expressed.txt",
# quote = F, row.names = F, col.names = F) #All expressed genes (background gene panel)
```

### Gene Ontology enrichment GORILLA (page 70)
```{r}
# webtool: https://cbl-gorilla.cs.technion.ac.il/
# using two unranked lists of genes
# Results: https://cbl-gorilla.cs.technion.ac.il/GOrilla/a0hgan88/GOResults.html

# Results show very interesting Enrichment for genes: Nefh, Nefl, Nefm
setwd(pwd2)
YMP_Nef_ens <- read.table("GOrilla/YMP_Nef.txt", header = TRUE, sep = "\t")
Nef_gene_ids <- meta_genes[meta_genes$ensembl_gene_id %in% YMP_Nef_ens$Ensembl.ID,]
neurofilament_expr_YMP <- avg_sc_expr_YMP[Nef_gene_ids$ensembl_gene_id,]
neurofilament_expr_YMP$mgi_symbol <- Nef_gene_ids$mgi_symbol

max_treatment_DEG_YMP <- setNames(colnames(avg_sc_expr_YMP)[apply(avg_sc_expr_YMP[DEG_YMP,], 1, which.max)],DEG_YMP)
avg_expr_DEG_list_YMP <- tapply(names(max_treatment_DEG_YMP), max_treatment_DEG_YMP, function(x) avg_sc_expr_YMP[x,])


rownames(avg_expr_DEG_list_YMP) %in% neurofilament_expr_YMP$ensembl_gene_id_version

neurofilament_list_YMP <- lapply(avg_expr_DEG_list_YMP, function(df) df[rownames(df) %in% neurofilament_expr_YMP$ensembl_gene_id_version,])


#pos_cell_list_aDR <- lapply(avg_expr_DEG_list_aDR, function(df) df[rownames(df) %in% cell_proc_meta$ensembl_gene_id_version,])

layout(matrix(1:2, nrow = 1, byrow = T))
par(mar=c(3,3,3,3))
boxplot(neurofilament_expr_YMP, main = paste0(treatment, " (", nrow(neurofilament_expr_YMP),")")) # wie stelle ich das gut dar?
#          ylim = c(-6e-16, 6e-16))

neurofilament_expr_YMP_long <- neurofilament_expr_YMP %>%
  gather(key = "treatment", value = "Expression", -mgi_symbol) # reshape the df
neurofilament_expr_YMP_long$Expression <- round(neurofilament_expr_YMP_long$Expression * 10e14, digits = 3)

grouping_plot <- ggplot(neurofilament_expr_YMP_long, aes(x = treatment, y = Expression, group = mgi_symbol, color = mgi_symbol)) +
  geom_line() +
  geom_point() +
  labs(title = "Process gene enrichment\nin YMP: Neurofilament",
       x = "Treatment",
       y = "Normalized expression") +
  theme_minimal() +
  theme(plot.title = element_text(hjust = 0.5))
grouping_plot
ggsave("Grouping_GO_Neurofilament.png",plot = grouping_plot, width = 3.5, height = 3, path = paste0(pwd2, "/plots"))

```







# Untrimmed data
## Import of data & comparison initial transcriptome analysis
### Import Gene expression quantification (p. 41)
```{r, warning=FALSE}
setwd(pwd)

samples <- list.files("rsem_genes_results_untrimmed")

expr <- sapply(samples, function(sample){
  file <- paste0("rsem_genes_results_untrimmed/", sample)
  #print(file)
  quant <- read.csv(file, sep="\t", header=T)
  tpm <- setNames(quant$TPM, quant$gene_id)
  return(tpm)
})
SRR_codes <- colnames(expr)
SRR_codes <- gsub("\\.genes\\.results", "", SRR_codes)
colnames(expr) <- SRR_codes
```



### Import metadata for mandatory (meta1) and optional (meta2) dataset (p. 41)
```{r, warning=FALSE}
setwd(pwd)

meta1 <- read.csv("SRR_Acc_List.txt", sep="\t", header=T) %>%
  inner_join(read.csv("PRJNA946653 Metadata.txt", header=T),
    by = c("Name" = "Run"),
    suffix = c("",".y")) 

meta2 <- read.csv("SRR_Acc_List.txt", sep="\t", header=T) %>%
  inner_join(read.csv("PRJNA875066 Metadata.txt", header=T),
    by = c("Name" = "Run"),
    suffix = c("",".y"))
# print(setdiff(colnames(meta1),colnames(meta2)))
meta1_filtered <- subset(meta1, select = c("Name", "experiment_group", "sex", "treatment")) #Compare this to dendrogram

meta1_filtered$sex <- as.factor(meta1_filtered$sex)
meta1_filtered$treatment <- as.factor(meta1_filtered$treatment)
meta1_filtered$experiment_group <- as.factor(meta1_filtered$experiment_group)

meta1_aDR <- meta1_filtered %>% filter(treatment == "AL" | treatment == "aDR" )
meta1_YMP <- meta1_filtered %>% filter(treatment == "YMP" | treatment == "PBS" )

summary(meta1_filtered$sex)
summary(meta1_filtered$treatment)
#expr <- expr[,meta$Name] #to make sure the columns of the expression matrix are in the same order as rows in the metadata
```


### Get more more information about genes with biomaRt (p. 42)
```{r}
#listDatasets(useEnsembl(biomart = "ensembl")) %>% filter(dataset == "mmusculus_gene_ensembl") #Is this the correct one?

library(biomaRt)

ensembl <- useEnsembl(biomart = "ensembl", dataset = "mmusculus_gene_ensembl")
meta_genes <- getBM(attributes = c("ensembl_gene_id",
                                   "ensembl_gene_id_version",
                                   "ensembl_transcript_id_version",
                                   "hgnc_symbol",
                                   "mgi_symbol", # appearantly necessary for DEseq
                                   "description",
                                   "chromosome_name",
                                   "start_position",
                                   "end_position",
                                   "strand"),
                    filters = "ensembl_gene_id_version",
                    values = rownames(expr),
                    mart = ensembl) %>%
right_join(data.frame(ensembl_gene_id_version = rownames(expr)), by = "ensembl_gene_id_version") %>% distinct(ensembl_gene_id_version, .keep_all = TRUE)

#expr_test <- expr[meta_genes$ensembl_transcript_id_version,]
```


### Comparison of transcriptome profiles 
#### Histograms, filtering unexpressed genes (p. 43-45)
```{r}
# dim(expr) # 56884 annotated genes and 15 samples
avg_expr <- rowMeans(expr)
#layout(matrix(1:2, nrow=1))
#hist(avg_expr)

hist(log10(avg_expr + 1))


ggplot(data.frame(avg_expr), aes(x=avg_expr)) +
  geom_histogram(bins = 50) +
  scale_x_continuous(breaks = c(0,1,10,100,1000,10000,100000), trans="log1p", expand=c(0,0)) +
  scale_y_continuous(breaks = c(0,1), expand=c(0,0), trans="log1p") +
  theme_minimal() #log transformation of the y axis, visualization purposes

num_det <- rowSums(expr > 0) #check in how many samples each gene is detected.
hist(num_det)
#head(num_det)


# get rid of genes that are unecpressed or extremely lowly expressed
expressed <- rowMeans(expr > 0) >= 0.5 | rowMeans(expr) >= 1

meta_genes$expressed <- expressed

head(expressed[meta_genes$expressed]) #expressed is a vector of booleans
summary(expressed[meta_genes$expressed])

#expr <- expr[which(expressed),] I want to keep the whole expr
dim(expr[meta_genes$expressed,]) # 25649    15 --> now only 25'649 genes left

avg_expr <- rowMeans(expr[meta_genes$expressed,]) #now after the filtering
#layout(matrix(1:2, nrow=1))
hist(avg_expr)
hist(log10(avg_expr + 1))
```
#### PCC & SCC (correlation coefficients, cluster dendorgrams, heatmap of samples) (p. 46)
```{r, fig.width=9}
#From here you need meta_genes
#meta_genes$expressed is expressed
#meta_genes$expressed

corr_pearson <- cor(log1p(expr[meta_genes$expressed,]))
corr_spearman <- cor(expr[meta_genes$expressed,], method = "spearman")

#pheatmap(corr_pearson)
plot <- pheatmap((corr_spearman), color = colorRampPalette(c((brewer.pal(n = 9, name = "YlGnBu"))))(100),  # custom color palette YlGnBu
         display_numbers = T,
         number_color = "black",
         number_format = "%.2f")
plot
#ggsave("spearman correlation of treated mice.svg",plot = plot, width = 9, height = 6, path = "/Users/david/Library/CloudStorage/OneDrive-Persönlich/Master ETH/Courses/11_Systems Genomics/project local repository/plots")



#Dendrograms
#hcl_pearson <- hclust(as.dist(1 - corr_pearson))
#hcl_spearman <- hclust(as.dist(1 - corr_spearman))
#layout(matrix(1:2,nrow=1))
#plot(hcl_pearson)
#plot(hcl_spearman)

```

#### PCA
```{r}
pca <- prcomp(log1p(t(expr[meta_genes$expressed,])), center = TRUE,scale.=TRUE)
variances <- pca$sdev^2

variance_proportion <- variances / sum(variances)
cumulative_variance <- cumsum(variance_proportion)


ggplot(data.frame(pca$x, meta1)) +
#geom_point(aes(x = PC1, y = PC2, color = Name, shape = treatment),size=5)
geom_point(aes(x = PC1, y = PC2, color = treatment),size=5)

# plot_ly(
#   x = pca$x[,1],
#   y = pca$x[,2],
#   type = "scatter",
#   text = colnames(expr), 
#   mode = "markers",
#   symbol = meta1_filtered$treatment,
#   marker = list(size = 10)
# ) %>%
#   layout(
#     xaxis = list(title = str_glue("PC1, variance: ", round(variance_proportion[1], 3))),
#     yaxis = list(title = str_glue("PC2, variance: ", round(variance_proportion[2], 3))),
#     hovermode = "closest"
#   )

```
##### batch effect correction? --> probbaly does not make sense as we do not have different layers
```{r}
library(sva)
expr_combat <- ComBat_seq(counts = expr,
                          batch = meta1$Name)

corr_spearman_combat <- cor(expr_combat[meta_genes$expressed,], method = "spearman")
hcl_spearman_combat <- hclust(as.dist(1 - corr_spearman_combat))
layout(matrix(1:2,nrow=1))
plot(hcl_spearman_combat, labels = meta1$Name)

pca_combat <- prcomp(log1p(t(expr_combat[meta_genes$expressed,])), center = TRUE, scale. = TRUE)
ggplot(data.frame(pca_combat$x, meta1)) +
  geom_point(aes(x = PC1, y = PC2, color = Layer, shape = Name), size = 5)
```

##### Highly variable genes identification (optional)
```{r, eval=FALSE}
#Highly variable genes identification (optional). Manual, We don't do it.
estimate_variability <- function(expr){
  means <- apply(expr, 1, mean)
  vars <- apply(expr, 1, var)
  cv2 <- vars / means^2
  
  
  minMeanForFit <- unname(median(means[which(cv2 > 0.3)]))
  useForFit <- means >= minMeanForFit
  fit <- glm.fit(x = cbind(a0 = 1, a1tilde = 1/means[useForFit]),
    y = cv2[useForFit],
    family = Gamma(link = "identity"))
  a0 <- unname(fit$coefficients["a0"])
  a1 <- unname(fit$coefficients["a1tilde"])
  
  
  xg <- exp(seq(min(log(means[means>0])), max(log(means)), length.out=1000))
  vfit <- a1/xg + a0
  df <- ncol(expr) - 1
  afit <- a1/means+a0
  varFitRatio <- vars/(afit*means^2)
  pval <- pchisq(varFitRatio*df,df=df,lower.tail=F)
  
  
  res <- data.frame(mean = means,
    var = vars,
    cv2 = cv2,
    useForFit = useForFit,
    pval = pval,
    padj = p.adjust(pval, method="BH"),
    row.names = rownames(expr))
  return(res)
}
```



## DESeq 2 (Differential expression Analysis)
### Getting isoforms information for DeSeq2
```{r, warning=FALSE}
ensembl <- useEnsembl(biomart = "ensembl", dataset = "mmusculus_gene_ensembl")
tx2gene <- getBM(attributes = c("ensembl_transcript_id_version","ensembl_gene_id_version"),
                     filters = "ensembl_gene_id_version",
                     values = rownames(expr),
                     mart = ensembl) %>%
  dplyr::select(ensembl_transcript_id_version, ensembl_gene_id_version)
```

```{r, warning=FALSE}
setwd(pwd)

samples <- list.files("rsem_isoforms_results_untrimmed") #Hier sollten isoforms sein! nicht dedup nextera!
filtered_samples <- samples[grep("\\.isoforms\\.results$",samples)]# Remove ".isoforms.results" from all column names
files_YMP <- file.path(paste0("rsem_isoforms_results_untrimmed/", filtered_samples[1:6]))
files_aDR <- file.path(paste0("rsem_isoforms_results_untrimmed/", filtered_samples[7:15]))
txi_YMP <- tximport(files_YMP, type = "rsem", tx2gene = tx2gene)
txi_aDR <- tximport(files_aDR, type = "rsem", tx2gene = tx2gene)

#expr[,15] <- data.frame(txi_aDR)[,9] # this code recovers the broken dataset from mouse 79
```

### DESeq aDR/AL (slide 57)
```{r}
dds_aDR <- DESeqDataSetFromTximport(txi_aDR, colData = meta1_aDR, design = ~ treatment) #the output we be in gene_ids!
#summary(dds_aDR)
rownames(dds_aDR) %>% length() #gene_id_versions
# txi_aDR, dds_aDR: 56884 genes

dds_filtered_aDR <- dds_aDR[intersect(meta_genes[meta_genes$expressed,2],rownames(dds_aDR)),]
rownames(dds_filtered_aDR) %>% length() #gene_id_versions
# sum(meta_genes$expressed): 22521; dds_filtered_aDR: 22521


dds_filtered_aDR <- DESeq(dds_filtered_aDR, test="LRT", reduced= ~ 1) #If only one covariate, use ~1

res_DESeq2_aDR <- results(dds_filtered_aDR) # 22521 genes, same amount
res_DESeq2_aDR <- data.frame(res_DESeq2_aDR) 
res_DESeq2_aDR$pvalue_BH <- p.adjust(res_DESeq2_aDR$pvalue, method="BH") #Argumentieren wieso BH correction
res_DESeq2_aDR$minuslog10_pvalue_BH <- -log10(res_DESeq2_aDR$pvalue_BH)

sum(is.na(res_DESeq2_aDR$log2FoldChange))
# the meta_genes dataset contains 6011 ensembl_gene_id_version that are not found in expr,
# because we filtered expr based on the criteria rowMeans(expr > 0) >= 0.5 | rowMeans(expr) >= 1

sum(is.na(res_DESeq2_aDR$pvalue)) # 6045
temp_mask = is.na(res_DESeq2_aDR$pvalue) & !is.na(res_DESeq2_aDR$log2FoldChange)
# res_DESeq2_aDR[temp_mask,] # However, for some reason DESeq2 fucks up the pvalue for these 47 genes
res_DESeq2_aDR_fitered <- res_DESeq2_aDR[!is.na(res_DESeq2_aDR$pvalue_BH),]
res_DESeq2_aDR_DEG <- res_DESeq2_aDR_fitered[res_DESeq2_aDR_fitered$pvalue_BH < 0.05,]#filter to only have significant data, include FC?
DEG_aDR <- rownames(res_DESeq2_aDR_DEG)

new_row <- c("untrimmed", "aDR", length(DEG_aDR))
DEG_comaprison <- rbind(DEG_comaprison, new_row)
```

#### Volcano Plot
```{r, fig.width=7, fig.height=7}
#plot(res_DESeq2_aDR$log2FoldChange, p.adjust(res_DESeq2_aDR$pvalue, method = "BH"))
#plot(res_DESeq2_aDR$log2FoldChange, -log10(p.adjust(res_DESeq2_aDR$pvalue)), method = "BH")

#BiocManager::install("EnhancedVolcano")
#library(EnhancedVolcano)

plot <- EnhancedVolcano(res_DESeq2_aDR_fitered,
                lab = rownames(res_DESeq2_aDR_fitered),
                labSize = 2,
                x = "log2FoldChange",
                y = "pvalue_BH",
                title = "DEG of DESeq2 aDR/AL",
                subtitle = "",
                #selectLab =,
                pCutoff = 0.05,
                FCcutoff = 1,
                pointSize = 3,
                xlim = c(-6,6),
                ylim = c(0, 10),
                drawConnectors = TRUE,
                directionConnectors = 'both',
                arrowheads = FALSE,
                boxedLabels = FALSE,
                legendPosition = "top")
plot
#ggsave("Volcano plot aDR_AL DESeq2.svg",plot = plot, width = 8, height = 8, path = "/Users/david/Library/CloudStorage/OneDrive-Persönlich/Master ETH/Courses/11_Systems Genomics/project local repository/plots")
```

#### Grouping of identified genes (p. 59)
```{r}
expr_aDR <- expr[,7:15]
avg_expr_aDR <- sapply(sort(unique(meta1_aDR$treatment)), function(treatment)
rowMeans(expr_aDR[,which(meta1_aDR$treatment == treatment)])) #V1: aDR, V2: AL
head(avg_expr_aDR)



#test, if this code is doing the right thing:
#rowMeans(expr_aDR[c("ENSMUSG00000000001.5","ENSMUSG00000000028.16"),which(meta1_aDR$treatment == "aDR")])

avg_expr_aDR <- data.frame(avg_expr_aDR)

```

```{r}
max_treatment_DEG_aDR <- setNames(colnames(avg_expr_aDR)[apply(avg_expr_aDR[DEG_aDR,], 1, which.max)],DEG_aDR)
avg_expr_DEG_list_aDR <- tapply(names(max_treatment_DEG_aDR), max_treatment_DEG_aDR, function(x) avg_expr_aDR[x,])
#avg_expr_DEG_list_aDR_df <- data.frame(avg_expr_DEG_list_aDR)
#avg_expr_DEG_list_aDR_df_up <- data.frame(avg_expr_DEG_list_aDR[1])
#avg_expr_DEG_list_aDR_df_up_scale <- scale(avg_expr_DEG_list_aDR_df_up)
#avg_expr_DEG_list_aDR_df_down <- data.frame(avg_expr_DEG_list_aDR[2])
test <- t(data.frame(avg_expr_DEG_list_aDR[1]))
#avg_expr_DEG_list_aDR_df[2,1]

scaled_expr_DEG_list_aDR <- lapply(avg_expr_DEG_list_aDR, function(x) t(scale(t(x),scale = FALSE))) # scaling converts all of the genes into one value instead of a normalized distribtion
layout(matrix(1:2, nrow = 1, byrow = T))
par(mar=c(3,3,3,3))
for(treatment in names(scaled_expr_DEG_list_aDR))
boxplot(scaled_expr_DEG_list_aDR[[treatment]],
main = paste0(treatment, " (", nrow(avg_expr_DEG_list_aDR[[treatment]]),")"))

```

#### hierarchical clustering of aDR GENES (MICE is below in a new chunk)
```{r}
#corr_DEG_mice <- cor(expr_aDR[DEG_aDR,], method = "spearman")
corr_DEG_genes <- cor(t(expr_aDR[DEG_aDR,]), method = "spearman")

library(pheatmap)
pheatmap(corr_DEG_genes)
hcl_DEG_genes <- hclust(as.dist(1 - corr_DEG_genes), method = "complete")
plot(hcl_DEG_genes, labels = FALSE)

#Virdis color palette
#install.packages("viridis")
library(viridis)
heatmap.2(corr_DEG_genes, Rowv = as.dendrogram(hcl_DEG_genes), Colv = as.dendrogram(hcl_DEG_genes),
trace = "none", scale = "none", labRow = NA, labCol = NA, col = viridis)

cl_DEG_genes <- cutree(hcl_DEG_genes, k = 3) #k can be adjusted
heatmap.2(corr_DEG_genes, Rowv = as.dendrogram(hcl_DEG_genes), Colv = as.dendrogram(hcl_DEG_genes),
trace = "none", scale = "none", labRow = NA, labCol = NA, col = viridis,
ColSideColors = rainbow(15)[cl_DEG_genes]) 
```
#### hierarchical clustering of aDR MICE (GENES is above)
```{r}
corr_DEG_mice <- cor(expr_aDR[DEG_aDR,], method = "spearman")
#corr_DEG_genes <- cor(t(expr_aDR[DEG_aDR,]), method = "spearman")

library(pheatmap)
pheatmap(corr_DEG_mice)
hcl_DEG_mice <- hclust(as.dist(1 - corr_DEG_mice), method = "complete")
plot(hcl_DEG_mice, labels = FALSE)

#Virdis color palette
#install.packages("viridis")
library(viridis)
heatmap.2(corr_DEG_mice, Rowv = as.dendrogram(hcl_DEG_mice), Colv = as.dendrogram(hcl_DEG_mice),
trace = "none", scale = "none", labRow = NA, labCol = NA, col = viridis)

cl_DEG_mice <- cutree(hcl_DEG_mice, k = 3) #k can be adjusted
heatmap.2(corr_DEG_mice, Rowv = as.dendrogram(hcl_DEG_mice), Colv = as.dendrogram(hcl_DEG_mice),
trace = "none", scale = "none", labRow = NA, labCol = NA, col = viridis,
ColSideColors = rainbow(15)[cl_DEG_mice]) 
```

#### 3-6 Making sense of the genes aDR (page 66)
```{r}
#aDR

#Convert list of DEG's into a format that you can upload into DAVID.
#Think of what cluster of genes we want to analyze!
#write.table(meta_genes[meta_genes$ensembl_gene_id_version %in% names(which(cl_DEG==2)), "ensembl_gene"], file = "genes_C2.txt",quote = F, row.names = F, col.names = F) #DEG's

#Convert DEG gene id versions to gene id (this is what DAVID wants)
# Remove everything after the dot
DEG_aDR_gene_id <- sub("\\..*", "", DEG_aDR)


write.table(DEG_aDR_gene_id,
file = "DEG_genes_aDR.txt",
quote = F, row.names = F, col.names = F) #DEG's, dies ist der Code der meiner Meinung nach Sinn macht.

write.table(meta_genes[meta_genes$expressed, "ensembl_gene_id"],
file = "genes_expressed.txt",
quote = F, row.names = F, col.names = F) #All expressed genes (background gene panel)

```

#### Enrichment analysis (page 70)
```{r}
DE_L4 <- DE_test(expr = expr[meta_genes$expressed,],
cond = meta$Layer == "L4",
ctrl = "FALSE",
covar = meta %>% dplyr::select(Individual)) %>%
tibble::rownames_to_column("gene")
scores <- setNames(sign(log(DE_L4$fc)) * (-log10(DE_L4$pval)),
setNames(meta_genes$ensembl_gene_id,
meta_genes$ensembl_gene_id_version)[DE_L4$gene])
scores_ordered <- sort(scores, decreasing=T)

library(msigdbr)
genesets_celltype <- msigdbr(species = "Homo sapiens", category = "C8")
genesets_celltype_list <- tapply(genesets_celltype$ensembl_gene, genesets_celltype$gs_name, list)
library(fgsea)
fgsea_kegg <- fgsea(pathways = genesets_celltype_list,
stats = scores_ordered,
minSize = 15,
maxSize = 500)

fgsea_kegg[order(NES,decreasing=T),][1:10,1:7]

plotEnrichment(genesets_celltype_list[["MANNO_MIDBRAIN_NEUROTYPES_HNBML5"]],
scores_ordered) + labs(title="hNbML5 GABAergic neurons")

fgsea_kegg[order(NES,decreasing=F),][1:10,1:7]



```




### DESeq YMP/PBS (slide 57)
```{r}
dds_YMP <- DESeqDataSetFromTximport(txi_YMP, colData = meta1_YMP, design = ~ treatment) #the output we be in gene_ids!
#summary(dds_aDR)
rownames(dds_YMP) %>% length() #gene_id_versions
# txi_aDR, dds_aDR: 56'884 genes

dds_filtered_YMP <- dds_YMP[intersect(meta_genes[meta_genes$expressed,2],rownames(dds_YMP)),]
rownames(dds_filtered_YMP) %>% length() #gene_id_versions
# sum(meta_genes$expressed): 22521; dds_filtered_aDR: 22521;


dds_filtered_YMP <- DESeq(dds_filtered_YMP, test="LRT", reduced= ~ 1) #If only one covariate, use ~1

res_DESeq2_YMP <- results(dds_filtered_YMP) # 22521 genes, same amount
res_DESeq2_YMP <- data.frame(res_DESeq2_YMP) 
res_DESeq2_YMP$pvalue_BH <- p.adjust(res_DESeq2_YMP$pvalue, method="BH") #Argumentieren wieso BH correction
res_DESeq2_YMP$minuslog10_pvalue_BH <- -log10(res_DESeq2_YMP$pvalue_BH)

sum(is.na(res_DESeq2_YMP$log2FoldChange))
# the meta_genes dataset contains 6693 ensembl_gene_id_version that are not found in expr,
# because we filtered expr based on the criteria rowMeans(expr > 0) >= 0.5 | rowMeans(expr) >= 1

sum(is.na(res_DESeq2_YMP$pvalue)) # 6693
temp_mask = is.na(res_DESeq2_YMP$pvalue) & !is.na(res_DESeq2_YMP$log2FoldChange)
# res_DESeq2_YMP[temp_mask,] # However, for some reason DESeq2 fucks up the pvalue for these 47 genes
res_DESeq2_YMP_fitered <- res_DESeq2_YMP[!is.na(res_DESeq2_YMP$pvalue_BH),]
res_DESeq2_YMP_DEG <- res_DESeq2_YMP_fitered[res_DESeq2_YMP_fitered$pvalue_BH < 0.05,]#filter to only have significant data, include FC?
DEG_YMP <- rownames(res_DESeq2_YMP_DEG)

new_row <- c("untrimmed", "YMP", length(DEG_YMP))
DEG_comaprison <- rbind(DEG_comaprison, new_row)
```

#### Volcano Plot
```{r, fig.width=7, fig.height=7}
#plot(res_DESeq2_aDR$log2FoldChange, p.adjust(res_DESeq2_aDR$pvalue, method = "BH"))
#plot(res_DESeq2_aDR$log2FoldChange, -log10(p.adjust(res_DESeq2_aDR$pvalue)), method = "BH")

#BiocManager::install("EnhancedVolcano")
library(EnhancedVolcano)

plot <- EnhancedVolcano(res_DESeq2_YMP_fitered,
                lab = rownames(res_DESeq2_YMP_fitered),
                labSize = 2,
                x = "log2FoldChange",
                y = "pvalue_BH",
                title = "DEG of DESeq2 YMP/PBS",
                subtitle = "",
                #selectLab =,
                pCutoff = 0.05,
                FCcutoff = 1,
                pointSize = 3,
                xlim = c(-6,6),
                ylim = c(0, 10),
                drawConnectors = TRUE,
                directionConnectors = 'both',
                arrowheads = FALSE,
                boxedLabels = FALSE,
                legendPosition = "top")
plot
#ggsave("Volcano plot aDR_AL DESeq2.svg",plot = plot, width = 8, height = 8, path = "/Users/david/Library/CloudStorage/OneDrive-Persönlich/Master ETH/Courses/11_Systems Genomics/project local repository/plots")
```


#### Grouping of identified genes (p. 59)
```{r}
expr_YMP <- expr[,1:6]
avg_expr_YMP <- sapply(sort(unique(meta1_YMP$treatment)), function(treatment)
rowMeans(expr_YMP[,which(meta1_YMP$treatment == treatment)])) #V1: YMP, V2: PBS
head(avg_expr_YMP)



#test, if this code is doing the right thing:
#rowMeans(expr_aDR[c("ENSMUSG00000000001.5","ENSMUSG00000000028.16"),which(meta1_aDR$treatment == "aDR")])

avg_expr_YMP <- data.frame(avg_expr_YMP)

```

```{r}
max_treatment_DEG_YMP <- setNames(colnames(avg_expr_YMP)[apply(avg_expr_YMP[DEG_YMP,], 1, which.max)],DEG_YMP)
avg_expr_DEG_list_YMP <- tapply(names(max_treatment_DEG_YMP), max_treatment_DEG_YMP, function(x) avg_expr_YMP[x,])
#avg_expr_DEG_list_aDR_df <- data.frame(avg_expr_DEG_list_aDR)
#avg_expr_DEG_list_aDR_df_up <- data.frame(avg_expr_DEG_list_aDR[1])
#avg_expr_DEG_list_aDR_df_up_scale <- scale(avg_expr_DEG_list_aDR_df_up)
#avg_expr_DEG_list_aDR_df_down <- data.frame(avg_expr_DEG_list_aDR[2])
test <- t(data.frame(avg_expr_DEG_list_YMP[1]))
#avg_expr_DEG_list_aDR_df[2,1]

scaled_expr_DEG_list_YMP <- lapply(avg_expr_DEG_list_YMP, function(x) t(scale(t(x),scale = FALSE))) # scaling converts all of the genes into one value instead of a normalized distribtion
layout(matrix(1:2, nrow = 1, byrow = T))
par(mar=c(3,3,3,3))
for(treatment in names(scaled_expr_DEG_list_YMP))
boxplot(scaled_expr_DEG_list_YMP[[treatment]],
main = paste0(treatment, " (", nrow(avg_expr_DEG_list_YMP[[treatment]]),")"))

```

#### hierarchical clustering of aDR GENES (MICE is below in a new chunk)
```{r}
#corr_DEG_mice <- cor(expr_aDR[DEG_aDR,], method = "spearman")
corr_DEG_genes <- cor(t(expr_YMP[DEG_YMP,]), method = "spearman")

library(pheatmap)
pheatmap(corr_DEG_genes)
hcl_DEG_genes <- hclust(as.dist(1 - corr_DEG_genes), method = "complete")
plot(hcl_DEG_genes, labels = FALSE)

#Virdis color palette
#install.packages("viridis")
library(viridis)
heatmap.2(corr_DEG_genes, Rowv = as.dendrogram(hcl_DEG_genes), Colv = as.dendrogram(hcl_DEG_genes),
trace = "none", scale = "none", labRow = NA, labCol = NA, col = viridis)

cl_DEG_genes <- cutree(hcl_DEG_genes, k = 3) #k can be adjusted
heatmap.2(corr_DEG_genes, Rowv = as.dendrogram(hcl_DEG_genes), Colv = as.dendrogram(hcl_DEG_genes),
trace = "none", scale = "none", labRow = NA, labCol = NA, col = viridis,
ColSideColors = rainbow(15)[cl_DEG_genes]) 
```
#### hierarchical clustering of aDR MICE (GENES is above)
```{r}
corr_DEG_mice <- cor(expr_YMP[DEG_YMP,], method = "spearman")
#corr_DEG_genes <- cor(t(expr_aDR[DEG_aDR,]), method = "spearman")

library(pheatmap)
pheatmap(corr_DEG_mice)
hcl_DEG_mice <- hclust(as.dist(1 - corr_DEG_mice), method = "complete")
plot(hcl_DEG_mice, labels = FALSE)

#Virdis color palette
#install.packages("viridis")
library(viridis)
heatmap.2(corr_DEG_mice, Rowv = as.dendrogram(hcl_DEG_mice), Colv = as.dendrogram(hcl_DEG_mice),
trace = "none", scale = "none", labRow = NA, labCol = NA, col = viridis)

cl_DEG_mice <- cutree(hcl_DEG_mice, k = 3) #k can be adjusted
heatmap.2(corr_DEG_mice, Rowv = as.dendrogram(hcl_DEG_mice), Colv = as.dendrogram(hcl_DEG_mice),
trace = "none", scale = "none", labRow = NA, labCol = NA, col = viridis,
ColSideColors = rainbow(15)[cl_DEG_mice]) 
```




## 3.7 Other analysis (page 72)










# Additional plots
```{r}
plot <- ggplot(DEG_comaprison, aes(x = `Rejuvenation approach`, y = DEGs, fill = `trim status`)) +
  geom_bar(stat = "identity", position = "dodge", colour="black", width = 0.7) +
  labs(title = "DEGs comparison untrimmed vs. trimmed", y = "Obtained DEGs") +
  theme_minimal() +
  scale_fill_manual(values = c("darkorchid","darkgreen")) +
  theme(legend.position = "bottom",
    legend.title = element_blank(),
    plot.title = element_text(hjust = 0.5),
    legend.text = element_text(size = 12)
    ) 

ggsave("DEGs_trimming_comparison.png",plot = plot, width = 9, height = 4, path = "/Users/valentin/Documents/ETH/Master Biotechnologie/Lectures/Sytems_Genomics/Local_git_hub_repository/plots")
```

