---
title: "Systems Genomics data analysis pipeline"
---

# install libraries
```{r}
#install.packages(c("tidyverse","ggrepel","BiocManager","pbapply","gplots","msigdbr"))
#BiocManager::install(c("biomaRt","sva","DESeq2","edgeR"))
```
```{r}
#BiocManager::install("tximport")
library(tximport)
```



# load libraries
```{r}
libraries = c("tidyverse","ggrepel","BiocManager","pbapply","gplots","msigdbr", "biomaRt","sva","DESeq2","edgeR", "dplyr", "biomaRt")
lapply(libraries, library, character.only=TRUE)
```



```{r}
#setwd("/Users/david/Library/CloudStorage/OneDrive-Persönlich/Master ETH/Courses/11_Systems Genomics/project local repository/data")
setwd("C:/Users/nicks/Desktop/Systems_Genomics/GitHub David/SystemsGenomics2023/data")

samples <- list.files("rsem_trimmed_dedup_nextera")
samples <- samples[seq(1, length(samples),by=2)]
#samples
expr <- sapply(samples, function(sample){
  file <- paste0("rsem_trimmed_dedup_nextera/", sample)
  #print(file)
  quant <- read.csv(file, sep="\t", header=T)
  #tpm <- setNames(quant$TPM, quant$gene_id)
  tpm <- setNames(quant$TPM, quant[,2])
  return(tpm)
})

```

```{r}
#setwd("/Users/david/Library/CloudStorage/OneDrive-Persönlich/Master ETH/Courses/11_Systems Genomics/project local repository/data")
setwd("C:/Users/nicks/Desktop/Systems_Genomics/GitHub David/SystemsGenomics2023/data")

meta2 <- read.csv("SRR_Acc_List.txt", sep="\t", header=T) %>%
  inner_join(read.csv("PRJNA875066 Metadata.txt", header=T),
    by = c("Name" = "Run"),
    suffix = c("",".y"))

meta1 <- read.csv("SRR_Acc_List.txt", sep="\t", header=T) %>%
  inner_join(read.csv("PRJNA946653 Metadata.txt", header=T),
    by = c("Name" = "Run"),
    suffix = c("",".y")) 

print(setdiff(colnames(meta1),colnames(meta2)))

meta1$sex <- as.factor(meta1$sex)
meta1$treatment <- as.factor(meta1$treatment)
summary(meta1$sex)
summary(meta1$treatment)

#expr <- expr[,meta$Name] #to make sure the columns of the expression matrix are in the same order as rows in the metadata
```

```{r}
#Conversion MUST (transcript) to MUSG (gene)

# Use the Ensembl dataset for Mus musculus
ensembl <- useEnsembl(biomart = "ensembl", dataset = "mmusculus_gene_ensembl")

# Specify the transcript IDs (ensm)
transcript_ids <- rownames(expr)

# Map transcript IDs to gene IDs (enmusg)
listAttributes(ensembl)
transcript_to_gene <- getBM(attributes = c("ensembl_transcript_id_version", "ensembl_gene_id_version"),
                            filters = "ensembl_transcript_id_version",
                            values = transcript_ids,
                            mart = ensembl)
transcript_to_gene$ensembl_transcript_id_version %>% unique() %>% length() #ca 150 000
transcript_to_gene$ensembl_gene_id_version %>% unique() %>% length()#ca  50 000

# Extract gene IDs (enmusg)
gene_ids_enmusg <- unique(transcript_to_gene$ensembl_gene_id_version)

# Print the result
print(gene_ids_enmusg)

```

#More information about genes with biomaRt (p. 42)
```{r}
listDatasets(useEnsembl(biomart = "ensembl")) %>% filter(dataset == "mmusculus_gene_ensembl") #Is this the correct one?
#str(expr)
#rownames(expr)
#head(rownames(expr),10)
#install.packages("biomaRt", dependencies=TRUE)
library(biomaRt)
listDatasets(ensembl)

ensembl <- useEnsembl(biomart = "ensembl", dataset = "mmusculus_gene_ensembl")


meta_genes <- getBM(attributes = c("ensembl_gene_id",
                                   "ensembl_gene_id_version",
                                   "ensembl_transcript_id_version",
                                   "hgnc_symbol",
                                   "description",
                                   "chromosome_name",
                                   "start_position",
                                   "end_position",
                                   "strand"),
                    filters = "ensembl_transcript_id_version",
                    values = rownames(expr),
                    mart = ensembl) %>%
right_join(data.frame(ensembl_transcript_id_version = rownames(expr)), by = "ensembl_transcript_id_version") %>% distinct(ensembl_transcript_id_version, .keep_all = TRUE)

#expr_test <- expr[meta_genes$ensembl_transcript_id_version,]
```
#Trying to get prtoein ID from ensembl
```{r}
setwd("C:/Users/nicks/Desktop/Systems_Genomics/GitHub David/SystemsGenomics2023/data")
expr_test <- sapply(samples, function(sample){
  file <- paste0("rsem_trimmed_dedup_nextera/", sample)
  #print(file)
  quant <- read.csv(file, sep="\t", header=T)
  #tpm <- setNames(quant$TPM, quant$gene_id)
  tpm <- setNames(quant$TPM, quant[,1])
  return(tpm)
})
expr_test %>% rownames() %>% unique() %>% length() #136 424, so only unique row names (protein id's)!
expr_test %>% nrow()

listAttributes(ensembl)
meta_genes_test <- getBM(attributes = c("ensembl_peptide_id", "external_gene_name", "description", "uniprotswissprot", "ensembl_gene_id", "ensembl_gene_id_version"),
                    filters = "uniprotswissprot",
                    values = rownames(expr_test),
                    mart = ensembl)

meta_genes_test$uniprotswissprot %>% unique() %>% length() #14374
```

#Histograms, filter unexpressed genes
```{r}
#print(meta1$Name)
#print(meta2$Name)
#print(samples)

dim(expr) #meaning 136424 annotated genes and 15 samples

avg_expr <- rowMeans(expr)
#layout(matrix(1:2, nrow=1))
hist(avg_expr)
hist(log10(avg_expr + 1))

library(ggplot2)
ggplot(data.frame(avg_expr), aes(x=avg_expr)) +
geom_histogram(bins = 50) +
scale_x_continuous(breaks = c(0,1,10,100,1000,10000,20000), trans="log1p", expand=c(0,0)) +
scale_y_continuous(breaks = c(0,1), expand=c(0,0), trans="log1p") +
theme_minimal() #log transformation of the y axis, visualization purposes

num_det <- rowSums(expr > 0) #check in how many samples each gene is detected.
hist(num_det)
#head(num_det)

expressed <- rowMeans(expr > 0) >= 0.5 | rowMeans(expr) >= 1
head(expressed) #expressed is a vector of booleans
summary(expressed)
expr <- expr[which(expressed),]
dim(expr)

avg_expr <- rowMeans(expr) #now after the filtering
#layout(matrix(1:2, nrow=1))
hist(avg_expr)
hist(log10(avg_expr + 1))


meta_genes$expressed <- expressed



```
#pairiwse correlation coefficients (p. 46)
```{r}
#From here you need meta_genes
#meta_genes$expressed is expressed
#meta_genes$expressed

corr_pearson <- cor(log1p(expr))
corr_spearman <- cor(expr, method = "spearman")

head(corr_pearson)
head(corr_spearman)
```
```{r}
#Dendrograms
hcl_pearson <- hclust(as.dist(1 - corr_pearson))
hcl_spearman <- hclust(as.dist(1 - corr_spearman))
layout(matrix(1:2,nrow=1))
plot(hcl_pearson)
plot(hcl_spearman)

```
```{r}
meta1_filtered <- subset(meta1, select = c("Name", "treatment")) #Compare this to dendrogram
meta1_filtered
```

```{r}
#PCA
pca <- prcomp(log1p(t(expr)), center = TRUE,scale.=TRUE)
eigs <- pca$sdev^2
plot(1:length(eigs), eigs)

ggplot(data.frame(pca$x, meta1)) +
#geom_point(aes(x = PC1, y = PC2, color = Name, shape = treatment),size=5)
  geom_point(aes(x = PC1, y = PC2, color = treatment),size=5)

```

```{r}
#Highly variable genes identification (optional). Manual, We don't do it.
estimate_variability <- function(expr){
  means <- apply(expr, 1, mean)
  vars <- apply(expr, 1, var)
  cv2 <- vars / means^2
  
  
  minMeanForFit <- unname(median(means[which(cv2 > 0.3)]))
  useForFit <- means >= minMeanForFit
  fit <- glm.fit(x = cbind(a0 = 1, a1tilde = 1/means[useForFit]),
    y = cv2[useForFit],
    family = Gamma(link = "identity"))
  a0 <- unname(fit$coefficients["a0"])
  a1 <- unname(fit$coefficients["a1tilde"])
  
  
  xg <- exp(seq(min(log(means[means>0])), max(log(means)), length.out=1000))
  vfit <- a1/xg + a0
  df <- ncol(expr) - 1
  afit <- a1/means+a0
  varFitRatio <- vars/(afit*means^2)
  pval <- pchisq(varFitRatio*df,df=df,lower.tail=F)
  
  
  res <- data.frame(mean = means,
    var = vars,
    cv2 = cv2,
    useForFit = useForFit,
    pval = pval,
    padj = p.adjust(pval, method="BH"),
    row.names = rownames(expr))
  return(res)
}
```
#DESeq (slide 57)
```{r}
library(tidyverse)
meta1_aDR <- meta1 %>% filter(treatment == "AL" | treatment == "aDR" )
meta1_YMP <- meta1 %>% filter(treatment == "YMP" | treatment == "PBS" )

library(biomaRt)
ensembl <- useEnsembl(biomart = "ensembl", dataset = "mmusculus_gene_ensembl")
tx2gene <- getBM(attributes = c("ensembl_transcript_id_version",
"ensembl_gene_id_version"),
filters = "ensembl_transcript_id_version",
values = rownames(expr),
mart = ensembl) %>%
dplyr::select(ensembl_transcript_id_version, ensembl_gene_id_version)

unique(tx2gene$ensembl_gene_id_version) %>% length()
unique(tx2gene$ensembl_transcript_id_version) %>% length()

setwd("C:/Users/nicks/Desktop/Systems_Genomics/GitHub David/SystemsGenomics2023/data")
library(tximport)
samples <- list.files("rsem_trimmed_dedup_nextera")
filtered_samples <- samples[grep("\\.isoforms\\.results$",samples)]
# Remove ".genes.results" from all column names
files <- file.path(paste0("rsem_trimmed_dedup_nextera/", filtered_samples))
files
txi <- tximport(files, type = "rsem", tx2gene = tx2gene)

library(DESeq2)
dds <- DESeqDataSetFromTximport(txi,colData = meta1, design = ~ treatment) #maybe include + sex?
summary(dds)
rownames(dds) %>% head() #gene_id_versions
rownames(dds) %>% length() #25095

dds_filtered <- dds[intersect(meta_genes[meta_genes$expressed,2], rownames(dds)),]
rownames(dds_filtered) %>% length() #16809



dds_filtered <- DESeq(dds_filtered, test="LRT", reduced= ~ 1) #If only one covariate, use ~1

res_DESeq2 <- results(dds_filtered)
plot(res_DESeq2$log2FoldChange, res_DESeq2$pvalue)


table(p.adjust(res_DESeq2$pvalue, method="BH") < 0.05) #Argumentieren wieso BH correction
      
#res_DE %>% filter(gene %in% rownames(res_DESeq2)) %>% pull(padj)< 0.1) #res_DE ist von der manuellen Analyse

#Plot only the significant log2FoldChanges, how?
```
#DESeq aDR/AL
```{r}
expr_aDR <- expr[,7:15] #only the columns with AL or aDR, 71-79
library(biomaRt)
ensembl <- useEnsembl(biomart = "ensembl", dataset = "mmusculus_gene_ensembl")
tx2gene_aDR <- getBM(attributes = c("ensembl_transcript_id_version",
"ensembl_gene_id_version"),
filters = "ensembl_transcript_id_version",
values = rownames(expr_aDR),
mart = ensembl) %>%
dplyr::select(ensembl_transcript_id_version, ensembl_gene_id_version)

setwd("C:/Users/nicks/Desktop/Systems_Genomics/GitHub David/SystemsGenomics2023/data")
library(tximport)
samples <- list.files("rsem_trimmed_dedup_nextera")
filtered_samples <- samples[grep("\\.isoforms\\.results$",samples)]
filtered_samples_aDR <- filtered_samples[7:15]
# Remove ".genes.results" from all column names
files <- file.path(paste0("rsem_trimmed_dedup_nextera/", filtered_samples_aDR))
files
txi_aDR <- tximport(files, type = "rsem", tx2gene = tx2gene_aDR)

library(DESeq2)
dds_aDR <- DESeqDataSetFromTximport(txi_aDR,colData = meta1_aDR, design = ~ treatment) #maybe include + sex?
summary(dds_aDR)
rownames(dds_aDR) %>% head() #gene_id_versions

dds_filtered_aDR <- dds[intersect(meta_genes[meta_genes$expressed,2],rownames(dds_aDR)),]
rownames(dds_filtered_aDR) %>% length() #gene_id_versions

dds_filtered_aDR <- DESeq(dds_filtered_aDR, test="LRT", reduced= ~ 1) #If only one covariate, use ~1

res_DESeq2_aDR <- results(dds_filtered_aDR)
plot(res_DESeq2_aDR$log2FoldChange, res_DESeq2_aDR$pvalue)

table(p.adjust(res_DESeq2_aDR$pvalue, method="BH") < 0.05) #Argumentieren wieso BH correction

plot(res_DESeq2_aDR$log2FoldChange, p.adjust(res_DESeq2_aDR$pvalue, method = "BH"))
```
#DESeq YMP

#Grouping of identified genes (p. 59)
```{r}
sum(is.na(res_DESeq2_aDR$pvalue))
res_DESeq2_aDR <- data.frame(res_DESeq2_aDR)
res_DESeq2_aDR$pvalue_BH <- p.adjust(res_DESeq2_aDR$pvalue, method="BH")

DEG_aDR <- res_DESeq2_aDR[res_DESeq2_aDR$pvalue_BH < 0.05,] #filter to only have significant data
DEG_aDR <- na.omit(DEG_aDR) 
DEG_aDR <- rownames(DEG_aDR) #extract rownames (names of gene id's)
DEG_aDR %>% head()

meta1_aDR$treatment %>% levels()

avg_expr_aDR <- sapply(sort(unique(meta1_aDR$treatment)), function(treatment)
rowMeans(expr_aDR[,which(meta1_aDR$treatment == treatment)])) #V1: aDR, V2: AL
head(avg_expr_aDR)

avg_expr_aDR <- data.frame(avg_expr_aDR)
avg_expr_aDR$ensembl_transcript_id_version <- rownames(avg_expr_aDR)
avg_expr_aDR$ensembl_transcript_id_version_mod <- sapply(strsplit(avg_expr_aDR$ensembl_transcript_id_version, ","), function(x) x[1])
#rownames(avg_expr_aDR) <- avg_expr_aDR$ensembl_transcript_id_version_mod
nrow(test)
length(unique(test$ensembl_gene_id_version))
#test <- na.omit(test)
colnames(transcript_to_gene)[1] <- "ensembl_transcript_id_version_mod"
left_join(avg_expr_aDR, transcript_to_gene) -> test #test: when 2 transcripts, we only keep one and match this to gene id -> no more NA is gene id
rownames(avg_expr_aDR) <- test$ensembl_gene_id_version

test$ensembl_gene_id_version
rownames(test) <- test$ensembl_gene_id_version
max_treatment_DEG <- setNames(colnames(avg_expr_aDR)[apply(test$ensembl_gene_id_version[DEG_aDR,], 1, which.max)],DEG_aDR)

max_treatment_DEG <- setNames(colnames(avg_expr_aDR)[apply(avg_expr_aDR[DEG_aDR,], 1, which.max)],DEG_aDR) 
#Problem: avg expr aDR has transcripts, DEG aDR has genes -> incompatible. Solution: transform avg expr aDR to genes
```

```{r}

table(max_layer_DEG)
avg_expr_DEG_list <- tapply(names(max_layer_DEG), max_layer_DEG, function(x) avg_expr[x,])
scaled_expr_DEG_list <- lapply(avg_expr_DEG_list, function(x) t(scale(t(x))))
layout(matrix(1:8, nrow = 2, byrow = T))
par(mar=c(3,3,3,3))
for(layer in names(scaled_expr_DEG_list))
boxplot(scaled_expr_DEG_list[[layer]],
main = paste0(layer, " (", nrow(scaled_expr_DEG_list[[layer]]),")"))

```




